matlab/OnlineDMD.m

```diff
@@ -0,0 +1,119 @@
+%       update(x,y), update when new snapshot pair (x,y) becomes available
+%                   Here, if the (discrete-time) dynamics are given by 
+%                   z(k) = f(z(k-1)), then (x,y) should be measurements 
+%                   correponding to consecutive states z(k-1) and z(k).
+%       computemodes(), compute and return DMD eigenvalues and DMD mdoes
+%
+% Authors: 
+%   Hao Zhang
+%   Clarence W. Rowley
+% 
+% Reference:
+% Hao Zhang, Clarence W. Rowley, Eric A. Deem, and Louis N. Cattafesta,
+% ``Online Dynamic Mode Decomposition for Time-varying Systems", 
+% in production, 2017. To be submitted for publication, available on arXiv.
+% 
+% Created:
+%   April 2017.
+%
+% To look up the documentation in the command window, type help OnlineDMD
+
+classdef OnlineDMD < handle
+    properties
+        n = 0;                      % state dimension
+        lambda = 1;                 % weighting factor
+        timestep = 0;               % number of snapshots processed
+        A;          % DMD matrix
+        P;          % matrix that contains information about past snapshots
+    end
+    
+    methods
+        function obj = OnlineDMD(n,lambda)
+            % Creat an object for online DMD
+            % Usage: odmd = OnlineDMD(n,lambda)
+            if nargin == 2
+                obj.n = n;
+                obj.lambda = lambda;
+                obj.A = zeros(n,n);
+                obj.P = zeros(n,n);
+            end
+        end
+        
+        function initialize(obj, Xq, Yq)
+            % Initialize OnlineDMD with q snapshot pairs stored in (Xq, Yq)
+            % Usage: odmd.initialize(Xq,Yq)
+            q = length(Xq(1,:));
+            if(obj.timestep == 0 && q>=obj.n)
+                sqrtlambda = sqrt(obj.lambda);
+                for i = 1:q
+                    Xq(:,i) = Xq(:,i)*sqrtlambda^(q-i);
+                    Yq(:,i) = Yq(:,i)*sqrtlambda^(q-i);
+                end
+                obj.A = Yq*pinv(Xq);
+                obj.P = inv(Xq*Xq')/obj.lambda;
+            end
+            obj.timestep = obj.timestep + q;
+        end
+        
+        function initializeghost(obj)
+            % Initialize online DMD with epsilon small (1e-15) ghost snapshot pairs before t=0
+            % Usage: odmd.initilizeghost()
+            epsilon = 1e-15;
+            alpha = 1.0/epsilon;
+            obj.A = randn(obj.n, obj.n);
+            obj.P = alpha*eye(obj.n);
+        end
+        
+        function update(obj, x, y)
+            % Update the DMD computation with a new pair of snapshots (x,y)
+            % Here, if the (discrete-time) dynamics are given by z(k) = f(z(k-1)), then (x,y)
+            % should be measurements correponding to consecutive states z(k-1) and z(k).
+            % Usage: odmd.update(x, y)
+            
+            % Compute gamma
+            gamma = 1/(1+x'*(obj.P*x));
```

Comment:
I have updated the matlab implementation. Now both matlab and python implementation are more efficient, matrix vector multiplication is computed beforehand.